src/coreclr/jit/codegenxarch.cpp

```diff
@@ -3453,18 +3409,180 @@ void CodeGen::genStructPutArgUnroll(GenTreePutArgStk* putArgNode)
 //
 void CodeGen::genStructPutArgRepMovs(GenTreePutArgStk* putArgNode)
 {
-    GenTree* srcAddr = putArgNode->gtGetOp1();
-    assert(srcAddr->TypeGet() == TYP_STRUCT);
+    GenTree* src = putArgNode->gtGetOp1();
+    assert(src->TypeGet() == TYP_STRUCT);
+    assert(!src->AsObj()->GetLayout()->HasGCPtr());
 
     // Make sure we got the arguments of the cpblk operation in the right registers, and that
-    // 'srcAddr' is contained as expected.
+    // 'src' is contained as expected.
     assert(putArgNode->gtRsvdRegs == (RBM_RDI | RBM_RCX | RBM_RSI));
-    assert(srcAddr->isContained());
+    assert(src->isContained());
 
     genConsumePutStructArgStk(putArgNode, REG_RDI, REG_RSI, REG_RCX);
     instGen(INS_r_movsb);
 }
 
+#ifdef TARGET_X86
+//------------------------------------------------------------------------
+// genStructPutArgPush: Generates code for passing a struct arg by value on stack using "push".
+//
+// Arguments:
+//     putArgNode  - the PutArgStk tree.
+//
+// Notes:
+//     Used only on x86, in two cases:
+//      - Structs 4, 8, or 12 bytes in size (less than XMM_REGSIZE_BYTES, multiple of TARGET_POINTER_SIZE).
+//      - Structs that contain GC pointers - they are guaranteed to be sized correctly by the VM.
+//
+void CodeGen::genStructPutArgPush(GenTreePutArgStk* putArgNode)
+{
+    // On x86, any struct that contains GC references must be stored to the stack using `push` instructions so
+    // that the emitter properly detects the need to update the method's GC information.
+    //
+    // Strictly speaking, it is only necessary to use "push" to store the GC references themselves, so for structs
+    // with large numbers of consecutive non-GC-ref-typed fields, we may be able to improve the code size in the
+    // future.
+    assert(m_pushStkArg);
+
+    GenTree* src     = putArgNode->Data();
+    GenTree* srcAddr = putArgNode->Data()->AsObj()->Addr();
+
+    regNumber  srcAddrReg   = srcAddr->GetRegNum();
+    const bool srcAddrInReg = srcAddrReg != REG_NA;
+
+    unsigned srcLclNum    = 0;
+    unsigned srcLclOffset = 0;
+    if (srcAddrInReg)
+    {
+        srcAddrReg = genConsumeReg(srcAddr);
+    }
+    else
+    {
+        assert(srcAddr->OperIsLocalAddr());
+
+        srcLclNum    = srcAddr->AsLclVarCommon()->GetLclNum();
+        srcLclOffset = srcAddr->AsLclVarCommon()->GetLclOffs();
+    }
+
+    ClassLayout*   layout   = src->AsObj()->GetLayout();
+    const unsigned byteSize = putArgNode->GetStackByteSize();
+    assert(byteSize % TARGET_POINTER_SIZE == 0);
```

Comment:
```suggestion
    assert((byteSize % TARGET_POINTER_SIZE == 0) && (byteSize  < XMM_REGSIZE_BYTES));
```

Reply:
Ah right. Is there a way to check something like `(either it is a structs with GC pointers) or (bytesize check that I suggested)`.

Reply:
That is certainly possible. Will add.